File: src/03_Hamming_Code/Main.java
import java.util.Scanner;

/**
 * This class provides methods to generate Hamming code, introduce errors, and
 * detect and correct them using the Hamming Code algorithm.
 */
class HammingCode {

    /**
     * Generates a parity bit for a given power (position of parity bit) based on the data.
     *
     * @param data  The data array which includes both data bits and parity bits
     * @param power The position of the parity bit in the Hamming code
     * @return The calculated parity bit (0 or 1)
     */
    static int generateParityBit(int data[], int power) {
        int parityBit = 0;
        for (int j = 0; j < data.length; j++) {
            if (data[j] != 2) { // 2 indicates a parity bit position
                int k = j + 1;
                String binaryPosition = Integer.toBinaryString(k);
                int z = ((Integer.parseInt(binaryPosition)) / ((int) Math.pow(10, power))) % 10;
                if (z == 1 && data[j] == 1) {
                    parityBit = (parityBit + 1) % 2; // Modulo-2 addition for parity bit
                }
            }
        }
        return parityBit;
    }

    /**
     * Generates a Hamming codeword by inserting parity bits into the original data.
     *
     * @param _data The original data array (data word)
     * @return The Hamming codeword (including data and parity bits)
     */
    static int[] generateCodeWord(int _data[]) {
        int dataLength = _data.length;
        int parityCount = 0, j = 0, k = 0;

        // Calculate how many parity bits are needed
        while (Math.pow(2, parityCount) < dataLength + parityCount + 1) {
            parityCount++;
        }

        int[] codeWord = new int[dataLength + parityCount]; // Codeword with parity bits

        // Initialize the codeword with data bits and '2' in parity bit positions
        for (int position = 1; position <= codeWord.length; position++) {
            if (Math.pow(2, j) == position) {
                codeWord[position - 1] = 2; // Insert placeholder for parity bit
                j++;
            } else {
                codeWord[position - 1] = _data[k++]; // Insert data bit
            }
        }

        // Calculate and insert the parity bits
        for (int position = 0; position < parityCount; position++) {
            codeWord[((int) Math.pow(2, position)) - 1] = generateParityBit(codeWord, position);
        }

        return codeWord;
    }

    /**
     * Checks the received Hamming code for errors and corrects them if any.
     *
     * @param codeword   The received codeword with potential errors
     * @param parityCount The number of parity bits in the codeword
     */
    static void receivedCode(int codeword[], int parityCount) {
        int[] parity = new int[parityCount]; // Array to store parity checks
        String syndrome = "";

        // Verify the parity bits
        for (int power = 0; power < parityCount; power++) {
            for (int i = 0; i < codeword.length; i++) {
                int k = i + 1;
                String s = Integer.toBinaryString(k);
                int bit = ((Integer.parseInt(s)) / ((int) Math.pow(10, power))) % 10;
                if (bit == 1 && codeword[i] == 1) {
                    parity[power] = (parity[power] + 1) % 2; // Modulo-2 addition for parity check
                }
            }
            syndrome = parity[power] + syndrome; // Build the syndrome string
        }

        // Check if there's an error
        int errorLocation = Integer.parseInt(syndrome, 2); // Convert syndrome to integer
        if (errorLocation != 0) {
            System.out.println("Error detected at position " + errorLocation);
            codeword[errorLocation - 1] = (codeword[errorLocation - 1] + 1) % 2; // Correct the error
            System.out.println("Error corrected. New codeword:");
            for (int i = codeword.length - 1; i >= 0; i--) {
                System.out.print(codeword[i]);
            }
            System.out.println();
        } else {
            System.out.println("No errors detected in the received codeword.");
        }

        // Extract the original data word
        System.out.println("The original data was:");
        int power = parityCount - 1;
        for (int i = codeword.length; i > 0; i--) {
            if (Math.pow(2, power) != i) {
                System.out.print(codeword[i - 1]);
            } else {
                power--;
            }
        }
        System.out.println();
    }
}

public class Main {

    public static void main(String[] args) {
        Scanner scan = new Scanner(System.in);

        // Sample data word (reverse sequence of actual data for easier calculation)
        int[] dataWord = new int[]{1, 0, 0, 1, 1, 0, 1};

        System.out.println("Original data word (in reverse):");
        for (int i = dataWord.length - 1; i >= 0; i--) {
            System.out.print(dataWord[i]);
        }
        System.out.println();

        // Generate the Hamming codeword
        int[] codeWord = HammingCode.generateCodeWord(dataWord);

        System.out.println("Generated Hamming codeword:");
        for (int i = codeWord.length - 1; i >= 0; i--) {
            System.out.print(codeWord[i]);
        }
        System.out.println();

        // Allow user to introduce an error in the codeword
        System.out.println("Enter the position of the bit to introduce an error (0 for no error):");
        int errorPosition = 0;
        try {
            errorPosition = scan.nextInt();
        } catch (Exception e) {
            System.out.println("Invalid input. Please enter a valid number.");
            return;
        }

        if (errorPosition != 0 && errorPosition <= codeWord.length) {
            codeWord[errorPosition - 1] = (codeWord[errorPosition - 1] + 1) % 2;
        } else if (errorPosition > codeWord.length) {
            System.out.println("Error: Invalid position. Exiting.");
            return;
        }

        // Display the modified codeword
        System.out.println("Sent codeword:");
        for (int i = codeWord.length - 1; i >= 0; i--) {
            System.out.print(codeWord[i]);
        }
        System.out.println();

        // Detect and correct errors
        HammingCode.receivedCode(codeWord, codeWord.length - dataWord.length);
    }
}

File: src/03_Hamming_Code/Main.py
"""
Hamming Code implementation in Python.
This script generates a Hamming codeword, allows introducing an error,
and detects/corrects it using the Hamming Code algorithm.
"""

import math


def generate_parity_bit(data, power):
    """
    Generate parity bit for a given power (position of parity bit).
    Args:
        data (list[int]): List containing data bits and parity placeholders.
        power (int): The position of the parity bit.
    Returns:
        int: The calculated parity bit (0 or 1).
    """
    parity_bit = 0
    for j in range(len(data)):
        if data[j] != 2:  # 2 indicates a parity placeholder
            k = j + 1
            binary_position = bin(k)[2:]  # binary string of index
            z = (int(binary_position) // (10 ** power)) % 10
            if z == 1 and data[j] == 1:
                parity_bit = (parity_bit + 1) % 2
    return parity_bit


def generate_codeword(data_word):
    """
    Generate the Hamming codeword by inserting parity bits.
    Args:
        data_word (list[int]): Original data bits.
    Returns:
        list[int]: Codeword including data and parity bits.
    """
    data_length = len(data_word)
    parity_count = 0

    # Find required number of parity bits
    while (2 ** parity_count) < (data_length + parity_count + 1):
        parity_count += 1

    codeword = [0] * (data_length + parity_count)
    j = 0
    k = 0

    # Insert data bits and parity placeholders (2)
    for position in range(1, len(codeword) + 1):
        if position == 2 ** j:
            codeword[position - 1] = 2
            j += 1
        else:
            codeword[position - 1] = data_word[k]
            k += 1

    # Calculate parity bits
    for position in range(parity_count):
        codeword[(2 ** position) - 1] = generate_parity_bit(codeword, position)

    return codeword


def received_code(codeword, parity_count):
    """
    Detect and correct errors in received Hamming code.
    Args:
        codeword (list[int]): Received codeword (possibly with error).
        parity_count (int): Number of parity bits.
    """
    parity = [0] * parity_count
    syndrome = ""

    # Verify parity bits
    for power in range(parity_count):
        for i in range(len(codeword)):
            k = i + 1
            binary_pos = bin(k)[2:]
            bit = (int(binary_pos) // (10 ** power)) % 10
            if bit == 1 and codeword[i] == 1:
                parity[power] = (parity[power] + 1) % 2
        syndrome = str(parity[power]) + syndrome

    error_location = int(syndrome, 2)
    if error_location != 0:
        print(f"\nError detected at position {error_location}")
        codeword[error_location - 1] = (codeword[error_location - 1] + 1) % 2
        print("Error corrected. New codeword:", "".join(map(str, reversed(codeword))))
    else:
        print("\nNo errors detected in the received codeword.")

    # Extract original data
    print("The original data was:", end=" ")
    power = parity_count - 1
    for i in range(len(codeword), 0, -1):
        if 2 ** power != i:
            print(codeword[i - 1], end="")
        else:
            power -= 1
    print()


def main():
    # Example data word (reversed for calculation, like Java version)
    data_word = [1, 0, 0, 1, 1, 0, 1]

    print("Original data word (in reverse):", "".join(map(str, reversed(data_word))))

    # Generate Hamming codeword
    codeword = generate_codeword(data_word)
    print("Generated Hamming codeword:", "".join(map(str, reversed(codeword))))

    # Introduce an error
    try:
        error_position = int(input("Enter the position to introduce error (0 for no error): "))
    except ValueError:
        print("Invalid input. Exiting.")
        return

    if error_position != 0 and error_position <= len(codeword):
        codeword[error_position - 1] = (codeword[error_position - 1] + 1) % 2
    elif error_position > len(codeword):
        print("Invalid position. Exiting.")
        return

    print("Sent codeword:", "".join(map(str, reversed(codeword))))

    # Detect and correct error
    received_code(codeword, len(codeword) - len(data_word))


if __name__ == "__main__":
    main()
